### PR TITLE
added rsat-core

### PR DESCRIPTION
**Pull request by jvanheld**
Bioconda requires reviews prior to merging pull-requests (PRs). To facilitate this, once your PR is passing tests and ready to be merged, please add the `please review & merge` label so other members of the bioconda community can have a look at your PR and either make suggestions or merge it.

* [ X ] I have read the [guidelines for bioconda recipes](https://bioconda.github.io/contributor/guidelines.html).
* [ X ] This PR adds a new recipe.
* [ X ] AFAIK, this recipe **is directly relevant to the biological sciences** (otherwise, please submit to the more general purpose [conda-forge channel](https://conda-forge.org/docs/)).
* [ ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

 * `@BiocondaBot please update` will cause the BiocondaBot to merge the master branch into a PR
 * `@BiocondaBot please add label` will add the `please review & merge` label.
 * `@BiocondaBot please fetch artifacts` will post links to packages and docker containers built by the CI system. You can use this to test packages locally before merging.

For members of the Bioconda project, the following command is also available:

 * `@BiocondaBot please merge` will cause packages/containers to be uploaded and a PR merged. Someone must approve a PR first! This has the benefit of not wasting CI build time required by manually merging PRs.

If you have questions, please post them in gitter or ping `@bioconda/core` in a comment (if you are not able to directly ping `@bioconda/core` then the bot will repost your comment and enable pinging).

### CODE REVIEW

**Comment by abretaud** [comment]
@jvanheld would it be possible to add a license file into the rsat archive? there's a linting error because of this (https://bioconda.github.io/contributor/linting.html#gpl-requires-license-distributed)

File: recipes/rsat-core/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  commands: 
+    - "rsat --help"   ## test the RSAT command
+    - "rsat oligo-analysis -h" ## Test a subcommand running a Perl script
+#    - "rsat implant-motif -h" ## Test a subcommand running a Perl script
+
+about:
+  home: http://rsat.eu
+  summary: Detection of regulatory signals in non-coding sequences.
+  license: AGPL-3
+  license_file: LICENSE.txt
```
Comment (bgruening) [inline recipes/rsat-core/meta.yaml]: @jvanheld can you add a DOI and the biotools id into an extra section please?

File: recipes/rsat-core/build.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (bgruening) [inline recipes/rsat-core/build.sh]: can you maybe add a shebang

File: recipes/rsat-core/build.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+RSAT_DEST="$PREFIX/opt/rsat/"
+mkdir -p "$RSAT_DEST"
+
+cp -a perl-scripts python-scripts makefiles bin/rsat share/rsat/rsat.yaml "$RSAT_DEST"
+
+# Build and dispatch compiled binaries
+cd contrib
+# for dbin in *
```
Comment (bgruening) [inline recipes/rsat-core/build.sh]: remove?

File: recipes/rsat-core/build.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+do
+    if [ -d "$dbin" ]; then
+        cd "$dbin"
+        make clean && make CC=$CC CXX=$CXX && cp "$dbin" "$PREFIX/bin"
+        cd ..
+    fi
+done
+
+
+# TODO: R packaging
```
Comment (bgruening) [inline recipes/rsat-core/build.sh]: still needed?

File: recipes/rsat-core/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    - vmatch
+
+    ## For the web server only?
+    #- perl-email-simple
+
+test:
+  commands: 
+    - "rsat --help"   ## test the RSAT command
+    - "rsat oligo-analysis -h" ## Test a subcommand running a Perl script
+#    - "rsat implant-motif -h" ## Test a subcommand running a Perl script
```
Comment (bgruening) [inline recipes/rsat-core/meta.yaml]: is this working now?

**Comment by bgruening** [comment]
> cp: cannot stat `bin/rsat': No such file or directory

**Comment by bgruening** [comment]
This seems to work now under Linux. Do you want to add the other modules back?

**Comment by abretaud** [comment]
@BiocondaBot please fetch artifacts

File: recipes/rsat-core/build.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+mkdir -p "$PREFIX/bin"
+mkdir -p "$PREFIX/share/rsat"
+
+cp -a LICENSE.txt perl-scripts python-scripts makefiles R-scripts "$RSAT_DEST"
+cp bin/rsat $PREFIX/bin/rsat
+cp share/rsat/rsat.yaml $PREFIX/share/rsat/rsat.yaml
+
+# Build and dispatch compiled binaries
+# cd contrib
+for dbin in info-gibbs
```
Comment (bgruening) [inline recipes/rsat-core/build.sh]: LGTM, but I guess @jvanheld want to include the other targets as well?

Comment (jvanheld) [reply recipes/rsat-core/build.sh]: Indeed, so far I commented almost everything i order to have something that would compile. I will now 

- compile the other C tools
- include them in the test
- add test for the other tools (Perl, python)
- compile the internal R package
- run a test for a workflow (peak-motifs) and a tool (matrix-clustering) that combine more or less everything 

If these tests pass, I guess I would consider that the suite is running fine :-)

File: recipes/rsat-core/build.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+cd contrib
+for dbin in info-gibbs count-words matrix-scan-quick retrieve-variation-seq variation-scan 
+do
+    cd "$dbin"
+    make clean && make CC=$CC CXX=$CXX && cp "$dbin" "$PREFIX/bin"
+    cd ..
+done
+cd ..
+
+## TEMPORARIY COMMENTED BECAUSE GENERATES ERROR
```
Comment (bgruening) [inline recipes/rsat-core/build.sh]: You need to add r-base and probably python as well to your requirements/host section.

**Comment by jvanheld** [comment]
Hi @bgruening , apparently there is a conflict between the default python  version and the python3 dependency of snakemake. Is it possible to include dependencies to both python2 and python3?

Thanks for your help

Jacques

**Comment by bgruening** [comment]
> Hi @bgruening , apparently there is a conflict between the default python version and the python3 dependency of snakemake. Is it possible to include dependencies to both python2 and python3?

No. You depend on python2? If you we should indicate this in the recipe and only build it for python2.

**Comment by jvanheld** [comment]
My mistake: I was confused by the message. All RSAT programs run on python 3 (I think, and if not we can adapt). Snakemake also depends on python3, so I dont understand why I had this error mentioning an incompatibility between python and snakemake. 

```

**Comment by jvanheld** [comment]
OK, all checks have passed. I included several tests 

```
test:
  commands: 
    - "rsat --help"   ## test the RSAT command
    - "rsat oligo-analysis -h" ## Test a subcommand running a Perl script
    - "rsat random-seq -l 100 -n 2" ## Test a perl script
    - "rsat random-seq -l 100 -n 2 | rsat purge-sequence" ## Check the vmatch dependency of purge-sequence
    - "rsat random-motif -l 10" ## Test a subcommand running a python script
    - "rsat feature-map -h" ## check the GD dependency for feature-map
    ## Check that the compiled executable are running (compiled C program, in the bin folder)
    - "rsat info-gibbs -h" 
    - "rsat count-words -h" 
    - "rsat matrix-scan-quick -h" 
```

They seem to have been taken into account in the container: 

```
23:18:07 BIOCONDA INFO (ERR) [Dec 23 23:18:07] STEP Run image [quay.io/biocontainers/rsat-core:2019.12.23--py37pl526r36hc99cbb1_0] with command [[rsat --help && rsat oligo-analysis -h && rsat random-seq -l 100 -n 2 && rsat random-seq -l 100 -n 2 | rsat purge-sequence && rsat random-motif -l 10 && rsat feature-map -h && rsat info-gibbs -h && rsat count-words -h && rsat matrix-scan-quick -h]]

```

For some reason, the result of the first test displayed is the only one that is displayed. 
At least I guess the other ones did not cause an error so I would consider that the recipe is functional.

**Comment by bgruening** [comment]
@BiocondaBot please fetch artifacts

**Comment by bgruening** [comment]
@jvanheld awesome. If you want please test the test-packages and/or the test container.

**Comment by bgruening** [comment]
> Yes, I will soon add the compilation of the R package

You want to do this in this PR or shoud we merge, so you have a Xmas present?

**Comment by jvanheld** [comment]
Hi @bgruening 

The idea of  xmas present is great, but I noticed that I still have a problem with the paths and I . My main command rsat calls subcommands which are in different paths. 

In the native rsat version, they are in 
rsat/bin ## the C and C++ compiled
rsat/perl-scripts
rsat/python-scripts

The rsat main command finds the list of subcommands in 

share/rsat/rsat.yaml

which indicates the sub-folder of each command relative to RSAT installation base directory
For example: 

```
  - command: "perl-scripts/random-genes"
  - command: "python-scripts/random-motif"
  - command: "bin/count-words"
...
```

The rsat command then searches the paths in share/rsat/
and from there I have relative links to the actual path of the perl-scripts, python)-scripts and bin directories. These paths will be different between the native RSAT and the conda distrib, but I guess this makes no problem since I can create the links.

I realised that I still need to add these links to enable the rsat command to find the paths. 

Then I seem to still have forgotten some perl dependency, I will add them and make a new trial. I guess I should still do it in PR ?

**Comment by bgruening** [comment]
>  I guess I should still do it in PR ?

Yes. Thanks.

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts; thanks a lot and merry Christmas, Mr Robot

**Comment by jvanheld** [comment]
Hi @abretaud  and @bgruening,

First of all,  merry xmas to both of you, and thanks for your continuous support. 
We almost did it for Christmas, but I think it will be for new year ;-)

I am progressing on RSAT packaging, but here I am a bit blocked: I need python 3.6 for seqlogo (and if possible for snakemake), but there seems to be an indirect dependency for python <2.8.0a0 (log below).

Is there a way to find what depends on python <2.8.0a0, or to circumvent the dependency conflict in another  way ?

Cheers,

Jacques
 


```
10:39:35 BIOCONDA INFO (OUT) Traceback (most recent call last):
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/environ.py", line 756, in get_install_actions
10:39:35 BIOCONDA INFO (OUT)     actions = install_actions(prefix, index, specs, force=True)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
10:39:35 BIOCONDA INFO (OUT)     return f(*args, **kwds)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/plan.py", line 473, in install_actions
10:39:35 BIOCONDA INFO (OUT)     txn = solver.solve_for_transaction(prune=prune, ignore_pinned=not pinned)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 107, in solve_for_transaction
10:39:35 BIOCONDA INFO (OUT)     force_remove, force_reinstall)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 145, in solve_for_diff
10:39:35 BIOCONDA INFO (OUT)     force_remove)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 242, in solve_final_state
10:39:35 BIOCONDA INFO (OUT)     ssc = self._run_sat(ssc)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
10:39:35 BIOCONDA INFO (OUT)     return f(*args, **kwds)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 510, in _run_sat
10:39:35 BIOCONDA INFO (OUT)     ssc.solution_precs = ssc.r.solve(tuple(final_environment_specs))
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
10:39:35 BIOCONDA INFO (OUT)     return f(*args, **kwds)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 1054, in solve
10:39:35 BIOCONDA INFO (OUT)     self.find_conflicts(specs)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 316, in find_conflicts
10:39:35 BIOCONDA INFO (OUT)     raise UnsatisfiableError(bad_deps)
10:39:35 BIOCONDA INFO (OUT) conda.exceptions.UnsatisfiableError: The following specifications were found to be in conflict:
10:39:35 BIOCONDA INFO (OUT)   - python[version='>=2.7,<2.8.0a0']
10:39:35 BIOCONDA INFO (OUT)   - seqlogo -> python[version='>=3']
10:39:35 BIOCONDA INFO (OUT) Use "conda search <package> --info" to see the dependencies for each package.
10:39:35 BIOCONDA INFO (OUT) 
10:39:35 BIOCONDA INFO (OUT) During handling of the above exception, another exception occurred:
10:39:35 BIOCONDA INFO (OUT) 
10:39:35 BIOCONDA INFO (OUT) Traceback (most recent call last):
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/bin/conda-build", line 11, in <module>
10:39:35 BIOCONDA INFO (OUT)     sys.exit(main())
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 445, in main
10:39:35 BIOCONDA INFO (OUT)     execute(sys.argv[1:])
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 436, in execute
10:39:35 BIOCONDA INFO (OUT)     verify=args.verify, variants=args.variants)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
10:39:35 BIOCONDA INFO (OUT)     notest=notest, need_source_download=need_source_download, variants=variants)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2343, in build_tree
10:39:35 BIOCONDA INFO (OUT)     notest=notest,
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1407, in build
10:39:35 BIOCONDA INFO (OUT)     create_build_envs(top_level_pkg, notest)
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1291, in create_build_envs
10:39:35 BIOCONDA INFO (OUT)     raise e
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1281, in create_build_envs
10:39:35 BIOCONDA INFO (OUT)     channel_urls=tuple(m.config.channel_urls))
10:39:35 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/environ.py", line 758, in get_install_actions
10:39:35 BIOCONDA INFO (OUT)     raise DependencyNeedsBuildingError(exc, subdir=subdir)
10:39:35 BIOCONDA INFO (OUT) conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-64: {"python[version='>=3']", "python[version='>=2.7,<2.8.0a0']"}
10:39:41 BIOCONDA ERROR COMMAND FAILED (exited with 1): docker run -t --net host --rm -v /tmp/tmp_kpsy9yq/build_script.bash:/opt/build_script.bash -v /home/circleci/project/miniconda/conda-bld/:/opt/host-conda-bld -v /home/circleci/project/recipes/rsat-core:/opt/recipe -e LANG=en_US.UTF-8 -e HOST_USER_ID=1001 bioconda/bioconda-utils-build-env:latest /bin/bash /opt/build_script.bash

10:39:41 BIOCONDA ERROR BUILD FAILED recipes/rsat-core

```

**Comment by jvanheld** [comment]
I temporarily inactivate seqloog and replace it by webglogo (seems to work)

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@ BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by bgruening** [comment]
@BiocondaBot please fetch artifacts

**Comment by bgruening** [comment]
@jvanheld finished?

**Comment by jvanheld** [comment]
Hi Bjorn, thanks for the follow-up. I am now revising each tool to make sure that it works with the reorganisation of the folders that we did for the conda packaging. This is also an opportunity for me to write a test for each command (in `makefiles/subcommand_tests.mk`). There are still some small problems to fix with some paths, but this revision will definitely be useful to consolidate the portability of the suite. 

I don't know whether I can already release it in the current state or if it is better to finish all the tests before having a first conda release.

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by jvanheld** [comment]
Hi @bgruening  and @abretaud 

Happy new year to you. 

I think the current version of RSAT package is already usable, even though I will need to improve some details for the management of genome data + fix dependencies for 3 Perl modules. I will try to treat these 3 Perl modules ASAP, and I think that we could then merge the conda package. 

Cheers,

Jacques

**Comment by jvanheld** [comment]
@BiocondaBot please fetch artifacts

**Comment by dpryan79** [comment]
@jvanheld Shall we merge this then?

**Comment by jvanheld** [comment]
Hi @dpryan79,

Yes, please do merge, this will already provide us with a first usable version. 

Thanks

Jacques

**Comment by dpryan79** [comment]
@bioconda-bot please merge

**Comment by bgruening** [comment]
BioContainer is available here: https://quay.io/repository/biocontainers/rsat-core?tab=tags

Thanks a lot @jvanheld!

**Pull request merged**